File: 2024/day_12/task.py
from collections import defaultdict
from time import time

start_time = time()

with open("/mnt/c/Users/aer/workspace/aoc/2024/day_12/input.txt") as file:
    data = file.read().split("\n")
    data = [list(line) for line in data]

plots = defaultdict(list)

visited = set()

for y,line in enumerate(data):
    for x,val in enumerate(line):
        plots[val].append((x,y))

def count_plots(plant_type):
    visited.add(plots[plant_type][0])
    regions = []
    count_1 = 0
    count_2 = 0

    for plant in plots[plant_type]:
        if plant in visited:
            continue
        plot = [plant]
        region = set(plot)

        while plot:        
            plant_x, plant_y = plot.pop()
            neighbors = [(plant_x+1, plant_y), (plant_x-1, plant_y), (plant_x, plant_y-1), (plant_x, plant_y+1)]
            for neighbor in neighbors:
                if len(data[0])-1 < neighbor[0] or neighbor[0] < 0 or len(data)-1 < neighbor[1] or neighbor[1] < 0:
                    continue                
                if (data[neighbor[1]][neighbor[0]] == plant_type and neighbor not in visited):
                    visited.add(neighbor)
                    plot.append(neighbor)
                    region.add(neighbor)
        
        regions.append(region)

           
        area = len(region)      
        count_region_1 = sum(1 
                    for 
                    (plant_x, plant_y) in region
                    for neighbor in 
                    [(plant_x+1, plant_y), (plant_x-1, plant_y), 
                     (plant_x, plant_y-1), (plant_x, plant_y+1)] 
                     if neighbor not in region)         
        count_1 += count_region_1 * area
             
    
    
        
        count_region = 0
        for x, y in region:
            if (x, y - 1) not in region and (
                (x - 1, y) not in region or (x - 1, y - 1) in region
            ):
                count_region += 1
            if (x, y + 1) not in region and (
                (x - 1, y) not in region or (x - 1, y + 1) in region
            ):
                count_region += 1
            if (x - 1, y) not in region and (
                (x, y - 1) not in region or (x - 1, y - 1) in region
            ):
                count_region += 1
            if (x + 1, y) not in region and (
                (x, y - 1) not in region or (x + 1, y - 1) in region
            ):
                count_region += 1
        count_2 += count_region * area
    
    return count_1, count_2

counts = [count_plots(i) for i in plots]

counts_1 = sum(c[0] for c in counts)
counts_2 = sum(c[1] for c in counts)

print(f"{counts_1}:{counts_2}: {time() - start_time:.3f}s")   

File: 2024/day_12/task.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Queue;
import java.util.Stack;


public class task {
    public static void main(String[] args) {
        Path path = Paths.get("./test.txt");        
        List<List<Character>> data = new ArrayList<>();
        
        try {
            Files.readAllLines(path).forEach((line) -> {
                var charrr = line.toCharArray();
                var chars = new ArrayList<Character>();

                for (char c : charrr) {
                    chars.add(c);
                }
                data.add(chars);
            });;
        } catch (IOException e) {
            System.err.println("ERROR");
        } 

        var plots = new HashMap<Character, List<int[]>>();

        for (int y = 0; y < data.size(); y++){
            for (int x = 0; x < data.getFirst().size(); x++){
                var plant_type = data.get(y).get(x);
                if (plots.containsKey(plant_type)) {
                    List<int[]> plants = plots.get(plant_type);
                    int[] point = {x, y};
                    plants.add(point);
                } else {
                    var plants = new ArrayList<int[]>();
                    int[] point = {x, y};
                    plants.add(point);
                    plots.put(plant_type, plants);
                }
                
            }
        }
        
        var visited = new HashSet<int[]>();

        int[] res = count_plots('R', visited, plots.get('R'));

        for (int i : res) {
            System.out.println(i);
        }
    }

    static boolean contains(Iterable<int[]> iter, int[] point) {
        for(int[] d: iter) {
            if (d[0] == point[0] && d[1] == point[1]) {
                return true;
            }
        }
        return false;
    }

    static int[] count_plots(Character plant_type, HashSet<int[]> visited, List<int[]> plots) {
        visited.add(plots.get(0));
        var count_1 = 0;
        var count_2 = 0;

        for (int[] plant: plots) {
            if (contains(visited, plant)) {
                continue;
            }
            var plot = new Stack<int[]>();
            plot.add(plant);
            var region = new HashSet<int[]>();
            region.add(plant);

            while (plot.size() > 0) {
                int[] current_plant = plot.pop();
                int[][] neighbors = {
                    {current_plant[0] + 1, current_plant[1]}, 
                    {current_plant[0] - 1, current_plant[1]}, 
                    {current_plant[0], current_plant[1] - 1}, 
                    {current_plant[0], current_plant[1] + 1}
                };

                for (int[] neighbor: neighbors) {
                    if (contains(plots, neighbor) && !contains(visited, neighbor)) {
                        visited.add(neighbor);
                        plot.add(neighbor);
                        region.add(neighbor);
                    }
                }
            }
            
            var area = region.size();
            var count_region_1 = 0;
            for (int[] current_plant: region) {
                System.out.println("(" + current_plant[0]+ ", " + current_plant[1] + ")");
                int[][] neighbors = {
                    {current_plant[0] + 1, current_plant[1]}, 
                    {current_plant[0] - 1, current_plant[1]}, 
                    {current_plant[0], current_plant[1] - 1}, 
                    {current_plant[0], current_plant[1] + 1}
                };
                for (int[] neighbor: neighbors) {
                    if (!contains(region, neighbor)) {
                        count_region_1++;
                    }
                }
            }
            System.out.println(count_region_1);
            System.out.println(area);
            count_1 += (count_region_1 * area);

        }
        
        int[] out = {count_1,2};
        return out;
    }
} 